Hardware property class

HazenBabcock opened the pull request:
Changes:
* Add a hardware property class `HWProperty` that is used for querying and setting camera properties.
* Change `ImagingDevice` to use the `HWProperty` class.
* Change `ImagingDevice` to keep track of the properties the user is interested in as determined by the values passed to `get_properties` and `set_properties`.

HazenBabcock committed:
Fix typo.

HazenBabcock committed:
Add hardware property class definition.

HazenBabcock committed:
Change ImagingDevice to use the HWProperties class.

campagnola commented on abstract/imaging_device.py:
```diff
@@ -1,14 +1,85 @@
+    """
+    def __init__(self, name = None, value = None, limits = None, values = None, writable = None, readable = None, **kwds):
+        super().__init(**kwds)
+
```
Should include a data type (float/int/bool/str)

HazenBabcock replied:
Agreed. There should also probably be a descriptor field as well.

HazenBabcock replied:
Another possibility would be to have sub-classes of `HWProperty` for each type of property (`HWPropertyFloat`, `HWPropertyInt`,..) . Not sure whether this is a better or worse idea than having a single class with a data type property.

campagnola replied:
Let's write it out and see. If the code gets messy (especially when considering type=array), then I'm not opposed to subclassing.

HazenBabcock replied:
I've added these changes in the most recent commit.

campagnola commented on abstract/imaging_device.py:
```diff
@@ -1,14 +1,85 @@
+        people use unreadable properties?
+        """
+
+    def is_valid(self, value):
```
Maybe return `None` in cases where validity can't be determined except by attempting to set the property.

HazenBabcock replied:
I agree, I had not thought of this but it is something we should have. Is there something that might be more specific than `None`? Maybe some kind of Exception? Though that doesn't seem like the right tool for the job. An `Unsure` object..

campagnola replied:
Good thought -- it's too easy for a bool vs None ambiguity to be missed by the user:
```
if not cam.is_valid(...):
    # we lane here regardless of whether is_valid returns False or None
```

campagnola commented on abstract/imaging_device.py:
```diff
@@ -70,21 +142,24 @@ def close():
         
     def get_properties(self, property_names):
-        """Return a dictionary of {'property_name': value} for the list of properties requested.
+        """Return a dictionary of {'property_name': HWProperty} for the list of properties requested.
```
Why return HWProperty here instead of the property value? (is this meant to be used differently from get_property_info?)

HazenBabcock replied:
I thought this function was going to return a list that only contained the names of all the properties, or maybe name and description. If we return all the properties to the user then they'd probably also have to be in the dictionary of properties that the user was interested in, since now they'll have all of them and we won't know which ones they really cared about. For some cameras this could be hundreds of properties, and might make it difficult to stay on top of which properties changed when the user called `set_properties()`.

campagnola replied:
I'm not really following here.. my intent was that this method is used to get the _values_ of specific properties requested by the user, like:
```
cam.get_properties(['exposure_time', 'binning_x', 'binning_y'])
# returns:  {
#    'exposure_time': 0.002,
#    'binning_x': 4,
#    'binning_y': 4,
# }
```
..and since the user has control over which values are being requested, I don't think it should be necessary for us to track `self.properties_used` ?

If the user wants to know what properties are available, then they'd use `get_property_info`.

HazenBabcock replied:
Sorry, my mistake, I crossed `get_properties` and `get_property_info` in my mind.

To answer the question that you originally asked. My thinking was to return `HWProperty` instead of value so that user could easily get all associated information about the property at the same time, for example the description, type, range, etc.. They'd also be allowed to change the `HWProperty` object (as it would be a copy) and pass it back to `set_properties`.

edbarnard commented:
Couple thoughts / questions:

1. Do we want hierarchical HW properties? For some more complex devices having a 100 flat properties would be unwieldy, but allowing for hierarchy adds complexity. One option would be to use a naming scheme that implies hierarchy ("x-axis/position")

2. For data types do we want to allow for arrays? For example a calibration curve etc.

campagnola commented:
> Do we want hierarchical HW properties?

In acq4 I have done simple things like `cam.set_property('region', (x, y, w, h))`, where `region` has 4 sub-properties that get assigned to. TBH, I think it only made things more difficult to write, and wasn't a significant benefit to the users of that API. .. but maybe you can suggest an example where it would be more useful? 

> For data types do we want to allow for arrays

+1 for this, but we need to think more carefully about how it will be used, since this is a mutable type unlike the others we've discussed so far. Maybe it's safe to always copy these arrays when they are assigned to a property?

HazenBabcock committed:
Add HWPropertyValueError() exception class. Add property type (ptype) and property descriptor (description) field to the HWProperty object. Organize keyword arguments, fields and methods so that they are all in the same order.

campagnola commented on abstract/imaging_device.py:
```diff
@@ -1,14 +1,121 @@
+    """
+    def __init__(self,
+                 name = None,
+                 value = None,
```
I'm thinking about this some more and wondering whether `value` does not belong here. I can definitely see using HWProperty as a way to describe the available properties and their limits, but it makes less sense to me as a way of getting/setting values on the device. I've generally found that I only request the property metadata when setting up user interfaces, and during actual acquisition I always just work with {property:value} dictionaries. There's also this issue that `get_properties` is redundant to `HWProperty.value`, except that the latter method cannot easily be made plural (I'd prefer there to be only one way to get/set property values).  Thoughts?

HazenBabcock replied:
In my mental model, this is addressing that a call to `set_properties` could change the allowed ranges, allowed values or even the value of one or more properties. I think there are two ways to handle how this flows back to the UI.

1. The device and the interface share the same `HWProperty` object so any updates happen "automatically". This I think is what you are saying regarding `HWProperty.value` and `get_properties` being redundant.
2. Each time properties change the device creates a updated list of `HWProperty` objects. In this case the `HWProperty.value` reflects the current state of the device, but changing it's value has no effect unless `set_properties` is called. The UI is responsible for updating it's model of the device state based on the new list of `HWProperty` objects, any old `HWProperty` objects that have the same names as those in the new list will be stale.

I'm in favor of (2) even though there will be a lot more churn of `HWProperty` objects because it breaks the implicit connection created by shared objects representing device state which I think will make debugging much harder.

campagnola replied:
Currently `set_properties` returns a dictionary of changed properties like `{'property': value}` and a list of the names of properties that might have their limits or other state changed, so I think that covers all of the information that needs to flow back to the UI, without the need for having `HWProperty.value`. One benefit of this approach is that it's light--we defer the effort of constructing those new HWProperty instances by just returning the names of properties that could have their limits changed.

HazenBabcock replied:
Okay, however we still need to have `HWProperty.value`, or else `get_properties` would need to change. The current proposal for `get_properties` is that it returns a dictionary of `HWProperty`.

campagnola replied:
I would suggest `get_properties` returns `{'prop_name': value, ...}`, as originally specified.

HazenBabcock replied:
Okay, so forget the `HWProperty` idea entirely? Not sure how you get a `HWProperty` object if it isn't return by either `get_properties` or `set_properties`.

campagnola replied:
Current description says `get_property_info` returns `{'prop_name': HWProperty}`

HazenBabcock replied:
As described it also says it would return information about all the properties supported by the device.

campagnola replied:
Ah, so maybe it should be `get_property_info(list_of_prop_names)` ?

HazenBabcock replied:
And an empty list? `None`? Would return a list of all the properties. Works for me.

HazenBabcock committed:
Remove 'value' from HWProperty object. Change 'get_properties' and 'set_properties' to return values / updated values instead of a HWProperty object with the new value. Change 'get_properties_info' to return requested HWProperty objects, or all of them if the 'property_names' argument is None.

HazenBabcock commented:
I changed `HWProperty` so that it no longer has a `value` attribute. I've also updated the `get_properties`, `get_properties_info` and `set_properties` methods.

campagnola reviewed (approved):
API looks good to me. There are a few code sections referencing `properties_used`; I'm not sure what these are for. (maybe they are left over from the get_properties/get_property_info confusion?)

I think the next thing I'd like to do is implement a subclass that wraps micromanager, and see how much work this is to plug in to acq4's hardware abstraction layer. Hopefully that will give some sense of whether there are any practical issues (and hopefully others will be motivated to try the same in their applications).

HazenBabcock commented:
Another option might be to just pick a random bit of hardware to write an interface for? Or maybe write an emulator, such as a camera emulator?
